### PR TITLE
Weiyao permission management create users permission

### PR DESCRIPTION
**Pull request by Weiyao-Li**
# Description
If the owner gives users like (Owners, Administrators, Core Team, Manager, Mentors, Volunteers) “create new users” access in “permission management, then this individual should see “user management” button, and be able to create new users, while at the same time, cannot use “delete”, “reset password”, “set final day”, “pause”, “request time off” these modification functionalities.

If the owner revokes anyone’s “create new users” access in “permission management”, then the person cannot see “user management” button.

The main changes are in the corresponding two files:
src/components/UserManagement/UserSearchPanel.jsx
src/components/UserManagement/UserTableData.jsx

## Related PRS (if any):
To test this frontend PR you need to checkout the development for backend

## Main changes explained:
- Disable modification buttons
- Give permission based on user's role

## How to test:
1. check into current branch
2. do `npm install` and `...` to run this PR locally
3. Clear site data/cache
4. log as volunteerA
5. go to dashboard→ Other links → You should NOT be able to see User Management you do not have “see all users” permission
6. log in as owner
7. go to dashboard→ Other links → Permission Management -> give your volunteerA account permission to create user -> logout
8. log as volunteerA
9. go to dashboard→ Other links → You SHOULD be able to see User Management
10. click User Management -> You should see all modification buttons disabled, except Create new user and sent setup link
11. You can login as other roles for testing

The above sample testing applies to Volunteer and Manager

This situation applies to Core Team, Mentor: If an user having the permission to “see all users” while do not have the permission to “create new user”, then the user can see “user management” button and therefore see all users, but cannot be able to create new user. The “create new user” + “modifications “ buttons will be disabled.

## Screenshots or videos of changes:
Detailed Video Walk-through with one example and edge cases
https://www.loom.com/share/48847f1043bd419c9366065d50da483c?sid=eba5bc92-83f0-450c-a2ae-34f00a1c354d

## Note:
Include the information the reviewers need to know.

### CODE REVIEW

**Review by nathanah — DISMISSED** [review]
Use `props.hasPermission()` to ensure that both personal and role permissions are checked. I DMed you the permissions documentation to show how to do that.

Remove the `isOwner` override. If somehow admins or owners do not have the necessary permissions, this will only cause confusion since any changes won't go through on the backend.

File: src/components/UserManagement/UserManagement.jsx [hunk 1]
```diff
@@ -84,6 +84,7 @@ class UserManagement extends React.PureComponent {
           <React.Fragment>
             {this.popupElements()}
             <UserSearchPanel
+              user={this.props.state.auth.user}
```
Comment (nathanah) [inline src/components/UserManagement/UserManagement.jsx:87]: Have you considered getting this through redux instead? It reduces prop drilling (making it easier to read/modify) and makes the code more consistent.

File: src/components/UserManagement/UserTableData.jsx [hunk 2]
```diff
@@ -25,18 +26,26 @@ const UserTableData = React.memo(props => {
...
   }, [props.isActive, props.resetLoading]);
 
+  const [isOwner,setIsOwner] = useState(false);
+  useEffect(() => {
+    if ( props.user?.role === 'Owner'|| props.user?.role === 'Administrator') {
+      setIsOwner(true)
+    } else {
+      setIsOwner(false)
+    }
+  },[])
```
Comment (nathanah) [inline src/components/UserManagement/UserTableData.jsx:36]:
```suggestion
```
Replace role-based logic with permission-based logic.

File: src/components/UserManagement/UserTableData.jsx [hunk 3]
```diff
@@ -111,6 +121,7 @@ const UserTableData = React.memo(props => {
           className={`btn btn-outline-primary btn-sm${
             props.timeOffRequests?.length > 0 ? ` time-off-request-btn-moved` : ''
           }`}
+          disabled={isOwner? false : true}
```
Comment (nathanah) [inline src/components/UserManagement/UserTableData.jsx:124]: It looks like this button is for requesting time off. Is that related to your PR??

File: src/components/UserManagement/UserTableData.jsx [hunk 4]
```diff
@@ -142,6 +153,7 @@ const UserTableData = React.memo(props => {
       <td>
         <button
           type="button"
+          disabled={isOwner? false : true}
```
Comment (nathanah) [inline src/components/UserManagement/UserTableData.jsx:156]:
```suggestion
```
Also unrelated to the "create new users" permission.

File: src/components/UserManagement/UserTableData.jsx [hunk 5]
```diff
@@ -181,6 +194,7 @@ const UserTableData = React.memo(props => {
           <span className="usermanagement-actions-cell">
             <button
               type="button"
+              disabled={isOwner? false : true}
```
Comment (nathanah) [inline src/components/UserManagement/UserTableData.jsx:197]:
```suggestion
```
This button is for deleting users.

File: src/components/UserManagement/UserTableData.jsx [hunk 6]
```diff
@@ -190,9 +204,9 @@ const UserTableData = React.memo(props => {
               {DELETE}
             </button>
           </span>
-          <span className="usermanagement-actions-cell">
-            <ResetPasswordButton authEmail={props.authEmail} user={props.user} isSmallButton />
-          </span>
+          {isOwner? <span className="usermanagement-actions-cell">
+             <ResetPasswordButton authEmail={props.authEmail} user={props.user} isSmallButton    />
+           </span>:null}
```
Comment (nathanah) [inline src/components/UserManagement/UserTableData.jsx:209]:
```suggestion
          <span className="usermanagement-actions-cell">
             <ResetPasswordButton authEmail={props.authEmail} user={props.user} isSmallButton    />
           </span>
```
Also not about creating users

File: src/components/UserManagement/__test__/UserSearchPanel.test.js
Comment (nathanah) [inline src/components/UserManagement/__test__/UserSearchPanel.test.js:1]: These tests should be fixed and not removed.

File: src/components/UserManagement/__tests__/UserSearchPanel.test.js
Comment (nathanah) [inline src/components/UserManagement/__tests__/UserSearchPanel.test.js:1]: These tests should be fixed and not removed.

File: src/components/UserManagement/UserTableData.jsx [hunk 7]
```diff
@@ -90,6 +99,7 @@ const UserTableData = React.memo(props => {
         <button
           type="button"
           className={`btn btn-outline-${props.isActive ? 'warning' : 'success'} btn-sm`}
+          disabled={isOwner? false : true}
```
Comment (nathanah) [inline src/components/UserManagement/UserTableData.jsx:102]:
```suggestion
```
Not related to creating users

File: src/components/UserManagement/UserTableData.jsx [hunk 8]
```diff
@@ -11,6 +11,7 @@ import { boxStyle } from 'styles';
 import { connect } from 'react-redux';
 import { formatDate } from 'utils/formatDate';
 import { cantUpdateDevAdminDetails } from 'utils/permissions';
+import { useSelector } from 'react-redux';
```
Comment (nathanah) [inline src/components/UserManagement/UserTableData.jsx:14]: At the bottom of this file, there's a `connect(null, { hasPermission })...` instead of having to import/use `useSelector` or `useDispatch` for everything, you can use mapStateToProps and mapDispatchToProps. Either way works, but connect() is a bit cleaner (all your redux state/functions in one place) and more in line with the coding practices used in this codebase.
https://react-redux.js.org/api/connect

File: src/components/UserManagement/UserSearchPanel.jsx [hunk 9]
```diff
@@ -1,18 +1,34 @@
...
+    } else {
+      setIsOwner(false)
+      const hasValue =props.user?.permissions.frontPermissions.some(value => value === 'postUserProfile');
+      if (hasValue) {
+        setHasRole(true);
+      } else {
+        setHasRole(false);
+      }
+    }
+  },[])
```
Comment (nathanah) [inline src/components/UserManagement/UserSearchPanel.jsx:23]:
```suggestion
  const canCreateUsers = props.hasPermission('postUserProfile');
```
It should look something like this.

File: src/components/UserManagement/UserSearchPanel.jsx [hunk 10]
```diff
@@ -1,18 +1,34 @@
...
   return (
     <div className="input-group mt-3" id="new_usermanagement">
-      <button type="button" className="btn btn-info mr-2" onClick={props.handleNewUserSetupPopup}>
+      <button disabled={isOwner ? false : !hasRole} type="button" className="btn btn-info mr-2"
+              onClick={props.handleNewUserSetupPopup}>
         {SEND_SETUP_LINK}
       </button>
       <button
+        disabled={isOwner ? false : !hasRole}
```
Comment (nathanah) [inline src/components/UserManagement/UserSearchPanel.jsx:31]:
```suggestion
        disabled={!canCreateUsers }
```
Remove role-based logic. Owners/admins should have the required permissions

File: src/components/UserManagement/UserSearchPanel.jsx [hunk 11]
```diff
@@ -1,18 +1,34 @@
...
+        setHasRole(true);
+      } else {
+        setHasRole(false);
+      }
+    }
+  },[])
   return (
     <div className="input-group mt-3" id="new_usermanagement">
-      <button type="button" className="btn btn-info mr-2" onClick={props.handleNewUserSetupPopup}>
+      <button disabled={isOwner ? false : !hasRole} type="button" className="btn btn-info mr-2"
```
Comment (nathanah) [inline src/components/UserManagement/UserSearchPanel.jsx:26]:
```suggestion
      <button disabled={canCreateUsers ? false : !hasRole} type="button" className="btn btn-info mr-2"
```

**Review by nathanah — DISMISSED** [review]
Needs to use permission logic instead of purely role-based logic.

File: src/components/UserManagement/UserTableData.jsx (hunk 4 above)
Comment (Weiyao-Li) [reply src/components/UserManagement/UserTableData.jsx]: > Also unrelated to the "create new users" permission.

I have updated the code. But we need to disable these.

Please check the feature description:
If the owner gives users like (Owners, Administrators, Core Team, Manager, Mentors, Volunteers) “create new users” access in “permission management, then this individual should see “user management” button, and be able to create new users, while at the same time, cannot use “delete”, “reset password”, “set final day”, “pause”, “request time off” these modification functionalities.

If the owner revokes anyone’s “create new users” access in “permission management”, then the person cannot see “user management” button.

File: src/components/UserManagement/UserTableData.jsx (hunk 5 above)
Comment (Weiyao-Li) [reply src/components/UserManagement/UserTableData.jsx]: > This button is for deleting users.

I have updated the code. But we need to disable these.

Please check the feature description:
If the owner gives users like (Owners, Administrators, Core Team, Manager, Mentors, Volunteers) “create new users” access in “permission management, then this individual should see “user management” button, and be able to create new users, while at the same time, cannot use “delete”, “reset password”, “set final day”, “pause”, “request time off” these modification functionalities.

If the owner revokes anyone’s “create new users” access in “permission management”, then the person cannot see “user management” button.

File: src/components/UserManagement/UserTableData.jsx (hunk 7 above)
Comment (Weiyao-Li) [reply src/components/UserManagement/UserTableData.jsx]: > Not related to creating users

I have updated the code. But we need to disable these.

Please check the feature description:
If the owner gives users like (Owners, Administrators, Core Team, Manager, Mentors, Volunteers) “create new users” access in “permission management, then this individual should see “user management” button, and be able to create new users, while at the same time, cannot use “delete”, “reset password”, “set final day”, “pause”, “request time off” these modification functionalities.

If the owner revokes anyone’s “create new users” access in “permission management”, then the person cannot see “user management” button.

File: src/components/UserManagement/UserTableData.jsx (hunk 4 above)
Comment (nathanah) [reply src/components/UserManagement/UserTableData.jsx]: "create new users" is one of the several permissions which grant the ability to see "user management". Other permissions pertaining to those separate actions (“delete”, “reset password”, “set final day”, “pause”, “request time off”) should grant access to the page and the ability to perform those actions, but should stay separate from the "create new users" permission.

File: src/components/UserManagement/__test__/UserSearchPanel.test.js [hunk 12]
```diff
@@ -2,27 +2,75 @@ import React from 'react';
...
+describe('user search panel not admin', () => {
   let onNewUserClick;
   let onSearch;
   let onActiveFilter;
+  let store;
   beforeEach(() => {
+    store = mockStore({
+      auth: nonJaeAccountMock,
+      userProfile: nonJaeAccountMock,
+      role: nonJaeAccountMock.role
```
Comment (nathanah) [inline src/components/UserManagement/__test__/UserSearchPanel.test.js:57]:
```suggestion
      role: rolesMock.role
```
Pretty sure the `role` here is actually an array of roles with their respective permissions used to determine which permissions roles have. I think it's a bit misleading that it's singular (the structure is actually `role: roles: {dictionary of roles[list of permissions for each role]}`)

This isn't the user's role, but about the permissions of roles in general.

File: src/components/UserManagement/__test__/UserSearchPanel.test.js [hunk 13]
```diff
@@ -2,27 +2,75 @@ import React from 'react';
...
+    email: 'devadmin@hgn.net'
+  },
+  firstName: 'Dev',
+  lastName: 'Admin',
+  weeklycommittedHours: 10,
+  email: 'devadmin@hgn.net',
+  role: {
+    roles:[]
+  }
+}
```
Comment (nathanah) [inline src/components/UserManagement/__test__/UserSearchPanel.test.js:46]:
```suggestion
```
It looks like this isn't used in this file.

File: src/components/UserManagement/__tests__/UserSearchPanel.test.js [hunk 14]
```diff
@@ -2,28 +2,71 @@ import React from 'react';
...
 describe('user search panel', () => {
   let onNewUserClick;
   let onSearch;
   let onActiveFilter;
+  let store;
   beforeEach(() => {
+    store = mockStore({
+      auth: ownerAccountMock,
+      userProfile: nonJaeAccountMock,
+      role: nonJaeAccountMock.role
```
Comment (nathanah) [inline src/components/UserManagement/__tests__/UserSearchPanel.test.js:51]:
```suggestion
      role: rolesMock.role
```
same as other role comment

File: src/components/UserManagement/__tests__/UserSearchPanel.test.js [hunk 15]
```diff
@@ -2,28 +2,71 @@ import React from 'react';
...
 
 describe('user search panel', () => {
   let onNewUserClick;
   let onSearch;
   let onActiveFilter;
+  let store;
   beforeEach(() => {
+    store = mockStore({
+      auth: ownerAccountMock,
+      userProfile: nonJaeAccountMock,
```
Comment (nathanah) [inline src/components/UserManagement/__tests__/UserSearchPanel.test.js:50]: Not sure why these are different.

File: src/components/UserManagement/UserTableData.jsx [hunk 16]
```diff
@@ -90,6 +95,7 @@ const UserTableData = React.memo(props => {
         <button
           type="button"
           className={`btn btn-outline-${props.isActive ? 'warning' : 'success'} btn-sm`}
+          disabled={!checkIsOwner()}
```
Comment (nathanah) [inline src/components/UserManagement/UserTableData.jsx:98]:
```suggestion
          disabled={!canChangeUserStatus}
```

File: src/components/UserManagement/UserTableData.jsx [hunk 17]
```diff
@@ -111,6 +117,7 @@ const UserTableData = React.memo(props => {
           className={`btn btn-outline-primary btn-sm${
             props.timeOffRequests?.length > 0 ? ` time-off-request-btn-moved` : ''
           }`}
+          disabled={!checkIsOwner()}
```
Comment (nathanah) [inline src/components/UserManagement/UserTableData.jsx:120]:
```suggestion
          disabled={!canChangeUserStatus} //or canRequestTimeOff???
```

File: src/components/UserManagement/UserTableData.jsx [hunk 18]
```diff
@@ -142,6 +149,7 @@ const UserTableData = React.memo(props => {
       <td>
         <button
           type="button"
+          disabled={!checkIsOwner()}
```
Comment (nathanah) [inline src/components/UserManagement/UserTableData.jsx:152]:
```suggestion
          disabled={!canChangeUserStatus} // or canSetFinalDay???
```

File: src/components/UserManagement/UserTableData.jsx [hunk 19]
```diff
@@ -190,9 +200,9 @@ const UserTableData = React.memo(props => {
               {DELETE}
             </button>
           </span>
-          <span className="usermanagement-actions-cell">
-            <ResetPasswordButton authEmail={props.authEmail} user={props.user} isSmallButton />
-          </span>
+          {checkIsOwner()? <span className="usermanagement-actions-cell">
```
Comment (nathanah) [inline src/components/UserManagement/UserTableData.jsx:203]:
```suggestion
          {canResetPassword? <span className="usermanagement-actions-cell">
```

**Review by nathanah — CHANGES_REQUESTED** [review]
